rest/src/main/java/io/syndesis/rest/v1/handler/setup/OAuthAppToCredentialsBridge.java

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * This component handles initializing and updating the CredentialProviderLocator
+ * from the DB state.
+ */
+@Component
+public class OAuthAppToCredentialsBridge {
+
+    private static final Logger LOG = LoggerFactory.getLogger(OAuthAppToCredentialsBridge.class);
+
+    private final OAuthAppHandler oAuthAppHandler;
+    private final CredentialProviderLocator locator;
+    private final Optional<EventBus> bus;
+    private ExecutorService executor;
+
+    public OAuthAppToCredentialsBridge(OAuthAppHandler oAuthAppHandler, CredentialProviderLocator locator, Optional<EventBus> bus) {
+        this.oAuthAppHandler = oAuthAppHandler;
+        this.locator = locator;
+        this.bus = bus;
+    }
+
+    @PostConstruct
+    public void init() {
+
+        if( bus.isPresent() ) {
+            executor = Executors.newSingleThreadExecutor();
+            bus.get().subscribe(getClass().getName(), getChangeEventSubscription());
+        }
+
+        // lets look at the app configs stored in the DB...
+        for (OAuthAppHandler.OAuthApp app : oAuthAppHandler.get()) {
+            registerCredentialProvider(app);
+        }
+    }
+
+    @PreDestroy
+    public void shutdown() {
+        if( bus.isPresent() ) {
+            bus.get().unsubscribe(getClass().getName());
+            executor.shutdown();
+        }
+    }
+
+    private EventBus.Subscription getChangeEventSubscription() {
+        return (event, data) -> {
+            // Never do anything that could block in this callback!
+            if (event!=null && "change-event".equals(event)) {
+                try {
+                    ChangeEvent changeEvent = Json.mapper().readValue(data, ChangeEvent.class);
+                    if (changeEvent != null) {
+                        changeEvent.getId().ifPresent(id -> {
+                            changeEvent.getKind()
+                                .map(Kind::from)
+                                .filter(k -> k == Kind.Connector)
+                                .ifPresent(k -> {
+                                    executor.execute(() -> onConnectorUpdated(id));
+                                });
+                        });
+                    }
+                } catch (IOException e) {
+                    LOG.error("Error while processing change-event " + data, e);
+                }
+            }
+        };
+    }
+
+    private void onConnectorUpdated(String connectorId) {
+        OAuthAppHandler.OAuthApp oAuthApp = null;
+        try {
+            oAuthApp = oAuthAppHandler.get(connectorId);
+        } catch (EntityNotFoundException e) {
+            // Not all connectors will be oauth apps..
+            return;
+        }
+        registerCredentialProvider(oAuthApp);
+    }
+
+    public void registerCredentialProvider(OAuthAppHandler.OAuthApp app) {
+        // is the app configured?
+        if( isSet(app.clientId) && isSet(app.clientSecret) ) {
+            if( "twitter".equals(app.id) ) {
```

Comment:
We could do it this way, or we could bring back the `CredentialProviderConfigurer`s that we chucked out in #439, and have the `id` property there, so they can serve as factories of `CredentialProvider`s, so that we don't have to explicitly check for each supported configuration in the future. Perhaps refine this in another PR.